packages/views.client.unstable/src/client.ts

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright 2024 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  type HostMessage,
+  META_TAG_HOST_ORIGIN,
+  type ParameterConfig,
+  type ParameterId,
+  type ViewMessage,
+} from "@osdk/views-api.unstable";
+
+type FoundryHostEventListener<CONFIG extends ParameterConfig> = (
+  event: MessageEvent<HostMessage<CONFIG>>,
+) => void;
+
+export interface IFoundryViewClient<CONFIG extends ParameterConfig> {
+  subscribe: (listener: FoundryHostEventListener<CONFIG>) => void;
+  unsubscribe: () => void;
+  sendMessage: (message: ViewMessage<CONFIG>) => void;
+}
+
+export class FoundryViewClient<CONFIG extends ParameterConfig>
+  implements IFoundryViewClient<CONFIG>
+{
+  private listener: FoundryHostEventListener<CONFIG> | undefined = undefined;
+  private parentWindow: Window;
+  private hostOrigin: string;
+
+  constructor() {
+    if (window.top == null) {
+      throw new Error("[FoundryViewClient] Must be run in an iframe");
+    }
+    this.parentWindow = window.top;
+    const metaTag = document.querySelector(
+      `meta[name="${META_TAG_HOST_ORIGIN}"]`,
+    );
+    if (metaTag == null) {
+      throw new Error(
+        "[FoundryViewClient] Missing host origin meta tag "
+          + META_TAG_HOST_ORIGIN,
+      );
+    }
```

Comment:
there's an argument to be made that maybe this shouldn't throw in development mode..? on the one hand, it's nice to be consistent in how the code runs whether its in production or not. on the other hand, it means you can't actually develop locally without using the developer mode that Foundry will provide for developing on these views

Reply:
Hard to say without thinking about what a disconnected dev/test setup would look like. We might want to provide a stubbed alternative to FoundryViewClient which mocks API calls. Otherwise you'd need a local mock HTML renderer which could provide this value anyway

Reply:
You definitely need to hard fail in development mode. You want people to realize whats happened easily.

You can also include a lot of extra info in console logs in development mode prior to throwing if you want. I also lean into the `tiny-invariant` library with the babel we have in repo. It rewrites these `invariant(window.top, "You must be running in an iframe")` to be:
```ts 
if (!window.top) {
  if ('production' !== process.env.NODE_ENV) {
    invariant(false, "You must be running in an iframe");
  } else {
    invariant(false);
  }
}
```

So this reduces down to almost nothing in production when the exception is mostly meaningless and non-actionable.

Reply:
If you take arguments that you want to default in a test environment I highly recommend NOT doing: `constructor(foo: Window = window.top)` as this breaks hard in some runtimes (I forget if node dies on this). Really you want to take an optional and then you can check the optional and assign it to the default in that case. You may also need to change to `globalThis.window?.top`. Definitely do additional research here.